core/kernel/tcg.c

```diff
@@ -0,0 +1,395 @@
+			TCG_EFI_SPEC_ID_EVENT_SPEC_VERSION_MAJOR_TPM2)
+		return TEE_ERROR_BAD_FORMAT;
+
+	if (spec_event->number_of_algorithms < 1) {
+		EMSG("specID Event: Number of algorithms incorrect");
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	alg_count = spec_event->number_of_algorithms;
+
+	if (tcg_get_pcr_info(&supported, &active, &pcr_count))
+		return TEE_ERROR_COMMUNICATION;
+
+	digest_list->count = 0;
+	/*
+	 * We have to take care that the sequence of algorithms that we record
+	 * in digest_list matches the sequence in eventlog.
+	 */
+	for (i = 0; i < alg_count; i++) {
+		hash_alg =
+		  get_unaligned_le16(&spec_event->digest_sizes[i].algorithm_id);
+
+		if (!(supported & tpm2_alg_to_tcg_mask(hash_alg))) {
+			EMSG("specID Event: Unsupported algorithm");
+			return TEE_ERROR_BAD_FORMAT;
+		}
+		digest_list->digests[digest_list->count++].hash_alg = hash_alg;
+
+		spec_active |= tpm2_alg_to_tcg_mask(hash_alg);
+	}
+
+	/*
+	 * TCG specification expects the event log to have hashes for all
+	 * active PCR's
+	 */
+	if (spec_active != active) {
+		/*
+		 * Previous stage bootloader should know all the active PCR's
+		 * and use them in the Eventlog.
+		 */
+		EMSG("specID Event: All active hash alg not present");
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	/*
+	 * the size of the spec event and placement of vendor_info_size
+	 * depends on supported algoriths
+	 */
+	spec_event_size =
+		offsetof(struct tcg_efi_spec_id_event, digest_sizes) +
+		alg_count * sizeof(spec_event->digest_sizes[0]);
+
+	if (*pos + spec_event_size >= log_size)
+		return TEE_ERROR_BAD_FORMAT;
+
+	vendor_sz = *(uint8_t *)((uintptr_t)buffer + *pos + spec_event_size);
+
+	spec_event_size += sizeof(vendor_sz) + vendor_sz;
+	*pos += spec_event_size;
+
+	if (get_unaligned_le32(&event_header->event_size) != spec_event_size) {
+		EMSG("specID event: header event size mismatch");
+		/* Right way to handle this can be to call SetActive PCR's */
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * tcg_parse_event() -  Parse the event in the eventlog
+ *
+ * @buffer:		Pointer to the start of the eventlog
+ * @log_size:		Size of the eventlog
+ * @offset:		[in] Offset of the event in the eventlog buffer
+ *			[out] Return offset of the next event in the buffer
+ * @digest_list:	list of digests in the event
+ * @pcr			Index of the PCR in the event
+ *
+ * Return:		status code
+ */
+static TEE_Result tcg_parse_event(void *buffer, uint32_t log_size,
+				  uint32_t *offset,
+				  struct tpml_digest_values *digest_list,
+				  uint32_t *pcr)
+{
+	struct tcg_pcr_event2 *event = NULL;
+	uint32_t count = 0, size = 0, event_size = 0;
+	uint32_t i = 0;
+	size_t pos = 0;
+
+	event_size = tcg_event_final_size(digest_list);
+	if (*offset >= log_size || *offset + event_size > log_size) {
+		EMSG("Event exceeds log size");
+		return TEE_ERROR_BAD_FORMAT;
+	}
+
+	event = (struct tcg_pcr_event2 *)((uintptr_t)buffer + *offset);
+	*pcr = get_unaligned_le32(&event->pcr_index);
```

Comment:
`struct tcg_pcr_event2` has the `__packed` attribute so you can access `event->pcr_index` directly.

By the way, if `struct tcg_pcr_event2` wouldn't had had the `__packed` attribute then would this `event` pointer be undefined.

Reply:
Offset in the buffer at which tcg_pcr_event2 is located is not necessarily aligned to a word boundary.

For clarity, I would like to keep the get_unaligned_le32 for the pcr_index